rosidl_runtime_cpp/include/rosidl_runtime_cpp/traits.hpp

```diff
@@ -15,11 +15,137 @@
 #ifndef ROSIDL_RUNTIME_CPP__TRAITS_HPP_
 #define ROSIDL_RUNTIME_CPP__TRAITS_HPP_
 
+#include <codecvt>
+#include <string>
 #include <type_traits>
 
 namespace rosidl_generator_traits
 {
 
+void value_to_yaml(bool value, std::ostream & out)
+{
+  out << (value ? "true" : "false");
+}
+
+void character_value_to_yaml(unsigned char value, std::ostream & out)
+{
+  auto flags = out.flags();
+  out << "0x" << std::hex << std::setw(2) << std::setfill('0') << +value;
+  out.flags(flags);
+}
+
+void character_value_to_yaml(char16_t value, std::ostream & out)
+{
+  auto flags = out.flags();
+  out << "\"\\u" << std::hex << std::setw(4) << std::setfill('0') << \
+    *reinterpret_cast<const int16_t *>(&value) << "\"";
```

Comment:
why the cast is needed?
isn't safer to `static_cast` between both?
```suggestion
    static_cast<const int16_t>(value) << "\"";
```

Reply:
> why the cast is needed?

`value` is a `char16_t` and the code wants to interpret it as a numeric value of type `int16_t`. Therefore I think `reinterpret_cast` is appropriate.

> isn't safer to `static_cast` between both?

What do you mean with "safer"?

Reply:
> What do you mean with "safer"?

`reinterpret_cast` allows to do more (and unsafer) conversions than `static_cast`.
By unsafer, I mean that it allows transmuting between everything.

If `static_cast` does the job, it should be preferred.

Reply:
I understand that. My point is that we *are* reinterpreting the data type. If you try it with a `static_cast` I am pretty certain it will result in a compile error because the types can't be statically casted between.

Reply:
All conversions between character types and integer types are allowed (see [is_integral](https://en.cppreference.com/w/cpp/types/is_integral), [Integral conversions](
 https://en.cppreference.com/w/cpp/language/implicit_conversion)). I have particularly tried that one, it works.

Considering that `char16_t` is binary equivalent to [uint_least16_t](https://en.cppreference.com/w/cpp/language/types), using `static_cast<const uint_least16_t>(value) << "\"";` would be the best choice (static cast from an unsigned integer to a signed integer was undefined until cpp20 if the number cannot be represented in the target type, does static_cast to `int16_t` might be an issue).

Reply:
> I have particularly tried that one, it works.

That is surprising. I did the same and got the following compiler error:

```
In file included from ws/build/rosidl_generator_cpp/rosidl_generator_cpp/rosidl_generator_cpp/msg/detail/empty__traits.hpp:9,
                 from ws/build/rosidl_generator_cpp/rosidl_generator_cpp/rosidl_generator_cpp/msg/empty.hpp:9,
                 from ws/src/ros2/rosidl/rosidl_generator_cpp/test/test_msg_datatype.cpp:16:
ws/install/rosidl_runtime_cpp/include/rosidl_runtime_cpp/traits.hpp: In function ‘void rosidl_generator_traits::character_value_to_yaml(char16_t, std::ostream&)’:
ws/install/rosidl_runtime_cpp/include/rosidl_runtime_cpp/traits.hpp:42:41: error: invalid static_cast from type ‘char16_t*’ to type ‘const int16_t*’ {aka ‘const short int*’}
   42 |     *static_cast<const int16_t *>(&value) << "\"";
      |                                         ^
make[2]: *** [CMakeFiles/test_msg_datatype.dir/build.make:63: CMakeFiles/test_msg_datatype.dir/test/test_msg_datatype.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:257: CMakeFiles/test_msg_datatype.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Reply:
because you cannot `static_cast` the pointer, but the value:
```cpp
static_cast<int16_t>(value);
```
It's still problematic to `static_cast` an unsigned integer to a `signed` one, so `uint_least16_t` should be preferred:
```cpp
static_cast<uint_least16_t>(value);
```

I don't think this is a blocker for merging the PR, but I think we should minimize `reinterpret_cast` usage when possible (which in this case is).

Reply:
Got it. Done in 5ffd92b1955315ffd0e6710f5bf68c54275b8b0c.